src/dune_rules/gen_rules.ml

```diff
@@ -146,6 +146,82 @@ end = struct
       | true ->
         let+ () = Mdx.gen_rules ~sctx ~dir ~scope ~expander mdx in
         empty_none)
+    | Melange mel ->
+      let melange_stanza_dir = dir in
+      let all_libs_compile_info =
+        let dune_version = Scope.project scope |> Dune_project.dune_version in
+        let pps = [] in
+        Lib.DB.resolve_user_written_deps_for_exes (Scope.libs scope)
+          [ (mel.loc, mel.target) ]
+          mel.libraries ~pps ~dune_version
+      in
+      let lib_cctx ~sctx ~obj_dir ~modules ~expander ~scope ~compile_info =
+        let flags = Ocaml_flags.empty in
+        let vimpl =
+          (* original impl in Lib_rules uses Virtual_rules.impl, will this break with virtual libs? *)
+          None
+        in
+        let modules = Vimpl.impl_modules vimpl modules in
+        let requires_compile = Lib.Compile.direct_requires compile_info in
+        let requires_link = Lib.Compile.requires_link compile_info in
+        let package = None in
+        let js_of_ocaml = None in
+        let melange = (melange_stanza_dir, mel.target) in
+        (* modes and pp are not passed, not sure if this will cause issues *)
+        Compilation_context.create () ~super_context:sctx ~expander ~scope
+          ~obj_dir ~modules ~flags ~requires_compile ~requires_link
+          ~opaque:Inherit_from_settings ~js_of_ocaml ~package ~melange ?vimpl
+      in
+      let rules (mel : Melange_stanza.t) ~sctx ~melange_stanza_dir ~scope
+          ~expander =
+        let open Memo.O in
+        let* libs = Lib.Compile.direct_requires all_libs_compile_info in
+        let* libs = Resolve.read_memo libs in
+        Memo.List.iter libs ~f:(fun lib ->
+            let lib_name = Lib.name lib in
+            let* lib, lib_compile_info =
+              Lib.DB.get_compile_info (Scope.libs scope) lib_name
+            in
+            let lib = Lib.Local.of_lib_exn lib in
+            let info = Lib.Local.info lib in
+            let lib_dir = Lib_info.src_dir info in
+            let obj_dir = Lib_info.obj_dir info in
+            let () =
+              if not (Path.Build.is_descendant lib_dir ~of_:melange_stanza_dir)
```

Comment:
Why do we need this check?

Reply:
Otherwise, the test `outside.t` will fail with:

```
Internal error, please report upstream including the contents of _build/log.
Description:
  ("[gen_rules] returned rules in a directory that is not a descendant of the directory it was called for",
  { dir = "default/inside"
  ; example =
      Rule
        { targets =
            { files = set { "default/app/.app.objs/b.ml.d" }; dirs = set {} }
        }
  })
```

It is unclear to me, what the final layout should be in that case?